buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -178,51 +178,60 @@ public static int hashCode(ByteBuf buffer) {
 
     /**
      * Returns {@code true} if and only if the two specified buffers are
-     * identical to each other as described in {@code ChannelBuffer#equals(Object)}.
-     * This method is useful when implementing a new buffer type.
+     * identical to each other for {@code len} bytes starting at {@code iLhs}
+     * index for the {@code lhs} buffer and {@code iRhs} index for the {@code rhs} buffer.
      */
-    public static boolean equals(ByteBuf bufferA, ByteBuf bufferB) {
-        final int aLen = bufferA.readableBytes();
-        if (aLen != bufferB.readableBytes()) {
+    public static boolean equals(ByteBuf lhs, int iLhs, ByteBuf rhs, int iRhs, int len) {
+        if (iLhs < 0 || iRhs < 0 || len < 0) {
+            throw new IllegalArgumentException("All indexes and lengths must be non-negative");
+        }
+        if (lhs.writerIndex() - len < iLhs || rhs.writerIndex() - len < iRhs) {
             return false;
         }
 
-        final int longCount = aLen >>> 3;
-        final int byteCount = aLen & 7;
+        final int longCount = len >>> 3;
+        final int byteCount = len & 7;
 
-        int aIndex = bufferA.readerIndex();
-        int bIndex = bufferB.readerIndex();
-
-        if (bufferA.order() == bufferB.order()) {
+        if (lhs.order() == rhs.order()) {
             for (int i = longCount; i > 0; i --) {
-                if (bufferA.getLong(aIndex) != bufferB.getLong(bIndex)) {
+                if (lhs.getLong(iLhs) != rhs.getLong(iRhs)) {
                     return false;
                 }
-                aIndex += 8;
-                bIndex += 8;
+                iLhs += 8;
+                iRhs += 8;
             }
         } else {
             for (int i = longCount; i > 0; i --) {
-                if (bufferA.getLong(aIndex) != swapLong(bufferB.getLong(bIndex))) {
+                if (lhs.getLong(iLhs) != swapLong(rhs.getLong(iRhs))) {
                     return false;
                 }
-                aIndex += 8;
-                bIndex += 8;
+                iLhs += 8;
+                iRhs += 8;
             }
         }
 
         for (int i = byteCount; i > 0; i --) {
-            if (bufferA.getByte(aIndex) != bufferB.getByte(bIndex)) {
+            if (lhs.getByte(iLhs) != rhs.getByte(iRhs)) {
                 return false;
             }
-            aIndex ++;
-            bIndex ++;
+            iLhs ++;
+            iRhs ++;
         }
 
         return true;
     }
 
     /**
+     * Returns {@code true} if and only if the two specified buffers are
+     * identical to each other as described in {@code ChannelBuffer#equals(Object)}.
+     * This method is useful when implementing a new buffer type.
+     */
+    public static boolean equals(ByteBuf bufferA, ByteBuf bufferB) {
+        return equals(bufferA, bufferA.readerIndex(), bufferB, bufferB.readerIndex(),
+                Math.max(bufferA.readableBytes(), bufferB.readableBytes()));
```

Comment:
I think we should return false directly if bufferA.readableBytes() != bufferB.readableBytes().